Update EEBUS to use version 0.6 of EEBUS libs

DerAndereAndi opened the pull request:
- Update to eebus-go v0.6.0, spine-go v0.6.0 and ship-go v0.5.1
- Add better support for VW ISO VAS capable chargers and EVs, to enable them to go to sleep

DerAndereAndi committed:
Update EEBUS to use version 0.5 of EEBUS libs

This updates the eebus charger implementation to use the latest version 0.5 implementations of cemd, eebus, spine and ship

DerAndereAndi commented:
This can also be used as a foundation to implement communication with a control box. But I recommend moving the eebus server then from charger to the server package.

I tested this with a software emulation of a wallbox so far. Will test with multiple wallboxes during the next week.

andig commented:
@DerAndereAndi ersetzt das dann https://github.com/evcc-io/evcc/pull/12407 oder ist das hier ein Zwischenschritt?

DerAndereAndi commented:
> @DerAndereAndi ersetzt das dann #12407 oder ist das hier ein Zwischenschritt?

Ja, das ersetzt deinen PR.

> Könntest Du die teilen oder finde ich die in https://github.com/enbility/cemd? Vielen Dank!

Nein, das geht leider nicht.

> Das wäre ein Thema dem wir uns widmen wollen wenn das Config UI released ist.

👍

DerAndereAndi committed:
Minor fix

DerAndereAndi committed:
Only use OSCEV use case in specific scenario

Only use OSCEV if the a wallbox with ISO15118-2 VW VAS is connected, and the EV is using this. This also enables full charging control and putting the EV to sleep.

All other current implementations of the usecase are basically wrong and are no different than using OPEV.

andig commented:
> Nein, das geht leider nicht.

Schade. Mir fehlt mangels Gerät jede Möglichkeit, hier etwas beizutragen ausser "trocken" coding.

Es wäre Klasse den PR fertig zu bekommen- aktuell ist aufgrund der geänderten Struktur keine (einfache) Aktualisierung der Dependencies mehr möglich.

andig commented on charger/eebus.go:
```diff
@@ -42,6 +43,7 @@ type EEBus struct {
 	minMaxG func() (minMax, error)
 
 	communicationStandard model.DeviceConfigurationKeyValueStringType
+	vasVW                 bool // wether the EVSE supports VW VAS with ISO15118-2
```
Was _ist_ VW VAS und warum hängt es vom Charger (statt vom Fahrzeug) ab?

DerAndereAndi replied:
Das sind Herstellerspezifische Erweiterungen des ISO15118-2 Protokolls. Das nutzt der Taycan mit der PMCC "Wallbox". Die Erweiterungen sind nicht öffentlich und damit nur mit bestimmten Kombinationen nutzbar.

andig replied:
> Diese spezielle Erweiterung ermöglicht die Übertragung des SoC und das Laden ohne dass das Auto selbst einen Ladebedarf hat weil das im Auto eingestellte Ladeziel schon erreicht wurde.

Ladeziel und -limit sind dann aber unabhängig von einander? Wir würden ja nicht ein ggf. im Fahrzeug gesetztes 80% Limit übersteuern wollen?

DerAndereAndi replied:
Die Fahrzeuge haben ein Min-SoC der gesetzt wird. Ein Max-SoC gibt es nur über Timer. Der Min-SoC ist per default auf 25%. Ein Ladelimit gibt es hier nicht wenn es eine ISO VAS Kommunikation gibt. Es gibt hier also nur Ladeziele, keine Limits.

Wenn man einen PMCC nutzt, ermöglicht das evcc in evcc das Ladeziel zu definieren, PV zu Laden ohne dass im Fahrzeug ein Timer gesetzt werden muss. Und gleichzeitig dass das Auto auch einschläft! Die Werte werden auch anders an das Auto übertragen als die Obligations.

Wenn man nur die "üblichen" Obligations nutzt und das Laden sperrt, dann schläft das Auto nicht ein. Denn es gibt bei ISO keine Definition/Mechanismus vom Protokoll es schlafen zu schicken.

Daher gibt es diese Unterscheidung für diesen speziellen Lademechanismus. Ich nutze den seit Jahren genau so und möchte dass das weiter so möglich ist. Der Vorteil von diesem geänderten Ansatz ist, dass das Auto auch wirklich einschlafen kann.

andig replied:
> Wenn man nur die "üblichen" Obligations nutzt und das Laden sperrt, dann schläft das Auto nicht ein. Denn es gibt bei ISO keine Definition/Mechanismus vom Protokoll es schlafen zu schicken.

Die Obligations bedeuten ja erstmal nicht, dass das Fahrzeug aktiv etwas tun muss/sollte. Ist das nicht ein Fehler der Fahrzeug-Firmware?

> Ich nutze den seit Jahren genau so und möchte dass das weiter so möglich ist. 

Spricht nichts gegen. Da ich es nirgendwo finden konnte: was heisst das Akronym `VAS`? Gibts dazu irgendeine Referenz? Vielen Dank!

DerAndereAndi replied:
Nein, es gibt hier keinen Fehler in der Firmware. Die Entwickler haben hier einfach weiter gedacht, wie man PV Laden direkt im Protokoll unterstützen kann und das umgesetzt. 

VAS = Vendor Added Services. Ist Teil der ISO 15118-2 Spezifikation.

andig commented:
Hi Andi, kann ich irgendwie helfen den PR zu finalisieren? Aufgrund der 0.x Versionen hängen wir sonst im Dependency Upgrade fest:

```
go: downloading golang.org/x/crypto/x509roots/fallback v0.0.0-20240604170348-d4e7c9cb6cb8
go: downloading golang.org/x/oauth2 v0.21.0
go: downloading github.com/itchyny/gojq v0.12.16
go: downloading github.com/go-playground/validator/v10 v10.21.0
go: downloading github.com/prometheus/common v0.54.0
go: downloading github.com/itchyny/timefmt-go v0.1.6
go: downloading modernc.org/sqlite v1.30.0
go: downloading github.com/prometheus/procfs v0.15.1
go: downloading google.golang.org/genproto v0.0.0-20240604185151-ef581f913117
go: downloading google.golang.org/genproto/googleapis/rpc v0.0.0-20240604185151-ef581f913117
go: downloading github.com/aws/aws-sdk-go v1.53.18
go: downloading github.com/insomniacslk/xjson v0.0.0-20240605162811-87a2577578b2
go: downloading modernc.org/libc v1.52.1
go: github.com/evcc-io/evcc imports
	github.com/evcc-io/evcc/cmd imports
	github.com/evcc-io/evcc/charger imports
	github.com/enbility/cemd/emobility: cannot find module providing package github.com/enbility/cemd/emobility
go: github.com/evcc-io/evcc imports
	github.com/evcc-io/evcc/cmd imports
	github.com/evcc-io/evcc/charger/eebus imports
	github.com/enbility/eebus-go/spine/model: cannot find module providing package github.com/enbility/eebus-go/spine/model
go: downloading modernc.org/ccgo/v4 v4.17.10
```

DerAndereAndi commented:
Ich kann erst nächste Woche wenn ich mein Auto wieder habe weiter machen.

Alternative für upgrades: du änderst die bisherige Nutzung von Versionsnummer auf den entsprechenden Commit.

DerAndereAndi commented:
go.mod direkt editieren mit:

```
github.com/enbility/eebus-go v0.0.0-20230521185917-1bae87b0ec27
github.com/enbility/cemd v0.0.0-20230426193025-28393246b3d0
```

sollte gehen. Bei CEMD evtl. das Datum auch auf `20230426183025` setzen.

andig commented:
Tut auch nicht:

```
go get -u
go: downloading github.com/enbility/cemd v0.0.0-20230426183025-28393246b3d0
go: github.com/evcc-io/evcc imports
	github.com/evcc-io/evcc/cmd imports
	github.com/evcc-io/evcc/charger imports
	github.com/enbility/cemd/emobility: cannot find module providing package github.com/enbility/cemd/emobility
go: github.com/evcc-io/evcc imports
	github.com/evcc-io/evcc/cmd imports
	github.com/evcc-io/evcc/charger/eebus imports
	github.com/enbility/eebus-go/spine/model: cannot find module providing package github.com/enbility/eebus-go/spine/model
```

Ich warte einfach ;)

andig commented:
Fixed by replacing the versions in https://github.com/evcc-io/evcc/commit/27d67d0d34ebce8c8d27cd532b586035607da8ea

DerAndereAndi committed:
Update EEBUS libraries and add fixes

- Updated to eebus-go v0.6.0, spine-go v0.6.0 and ship-go v0.5.1
- Various updates and fixes

DerAndereAndi commented:
Tested with two EEBUS compatible chargers and works great. It even seems to improve the crashing issues of the Elli Charger Connect and Pro wallboxes to some extend for me.

andig committed:
Merge branch 'master' into feature/eebus05

andig commented:
Ich hatte die use cases in https://github.com/evcc-io/evcc/pull/12407 noch (für meinen Geschmack) minimal eleganter erzeugt:

```go 
	// create use cases
	service := c.cem.Service
	c.uc = &Usecases{
		EvseCC: ucevsecc.NewUCEVSECC(service, c.evseHandler),
		EvCC:   ucevcc.NewUCEVCC(service, c.evHandler),
		EvCem:  ucevcem.NewUCEVCEM(service, c.evHandler),
		OpEV:   ucopev.NewUCOPEV(service, c.evHandler),
		EvSoc:  ucevsoc.NewUCEVSOC(service, c.evHandler),
	}

	// register use cases
	for _, uc := range []cemdapi.UseCaseInterface{
		c.uc.EvseCC,
		c.uc.EvCC, c.uc.EvCem, c.uc.OpEV, c.uc.EvSoc,
	} {
		c.cem.AddUseCase(uc)
	}
```

DerAndereAndi committed:
Improve code style per suggestion

andig committed:
Minor optimisation

andig commented:
Klasse- hab noch ein paar Kleinigkeiten geändert und die Cert-Funktionen raus gezogen.

andig committed:
Re-add IP for compatibility

andig commented:
Meinst Du das kann ins Nightly oder lieber bis nach Release warten?

DerAndereAndi commented:
Wenn du kurzfristig ein Release planst, lieber danach

andig commented:
Klasse- dann rein damit. Es steht grad nix Kritisches an.

andig commented on charger/eebus.go:
```diff
@@ -293,17 +336,37 @@ func (c *EEBus) Enabled() (bool, error) {
+	limits, err := c.uc.OpEV.LoadControlLimits(evEntity)
 	if err != nil {
-		// there are no overload protection limits available, e.g. because the data was not received yet
+		// there are limits available, e.g. because the data was not received yet
```
@DerAndereAndi der Kommentar sollte "no limits" heißen? Oder ist die Bedingung falsch?

DerAndereAndi replied:
Den Kommentar gibt es im aktuellen Code nicht und die Stelle greift nur bei VAS => Taycan. Das läuft sauber.

andig replied:
Arg, falsche Stelle. Den gibts immer noch:

```go
	limits, err := c.uc.OpEV.LoadControlLimits(evEntity)
	if err != nil {
		// there are limits available, e.g. because the data was not received yet
		c.log.DEBUG.Println("!! OpEV.LoadControlLimits error", err)
		return true, nil
	}
```

andig replied:
An der Stelle hier falsch?

DerAndereAndi replied:
Glaube ich kaum, denn Limits sind im Ablauf ja da und gesetzt.

andig replied:
Mir gehts nur um den Kommentar- der müsste heissen "no limits"? Dann würde ich das lokal korrigieren.
